src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperFactsService.cs

```diff
@@ -15,5 +16,51 @@ public abstract class TagHelperFactsService
         public abstract IReadOnlyList<TagHelperDescriptor> GetTagHelpersGivenTag(TagHelperDocumentContext documentContext, string tagName, string parentTag);
 
         public abstract IReadOnlyList<TagHelperDescriptor> GetTagHelpersGivenParent(TagHelperDocumentContext documentContext, string parentTag);
+
+        // Internal for testing
+        internal virtual IEnumerable<KeyValuePair<string, string>> StringifyAttributes(SyntaxList<RazorSyntaxNode> attributes)
+        {
+            var stringifiedAttributes = new List<KeyValuePair<string, string>>();
```

Comment:
Any reason to move the implementation up to the base class? Can this go to `DefaultTagHelperFactsService`?

Reply:
According to @NTaylorMullen `TagHelperFactsService` is a class that WTE has dependencies on, so I can't make any breaking changes to it such as the inclusion of new abstract methods. Virtual methods are OK though. I suppose we could have the `StringifyAttributes` on TagHelperFactsService throw NotImplementedException, but that just feels like abstract classes with extra steps.

Reply:
>  I suppose we could have the StringifyAttributes on TagHelperFactsService throw NotImplementedException, but that just feels like abstract classes with extra steps.

I still prefer doing that over having the implementation in an abstract class. It feels right for the actual implementation to reside on the class implementation rather than the abstract base class unless it is static. Thoughts? @ryanbrandenburg @NTaylorMullen

Reply:
Ya i agree with lifting the method impl into the implementation type